tripal/tripal
# Overhaul of content creation for prop-based types.

Issue #427 

## Type(s) of Change(s)


## Description


This PR was initiated because of a major problem with content type creation.  If a user created a content type that was specified using a property it would work fine for publishing those content types, but not for inserting new records via the GUI.  For example, if you wanted to create a 'genome_assembly' content type where the type of assembly is defined in the analysisprop table you could publish existing records but not add new ones.   There were two problems 1)  the code for adding new entities never added the appropriate prop record.  2) For tables such as the feature table that already have a type_id, but that type_id is not used to differentiate the content type, the insertion would fail (e.g. feature table entities). This is because the type_id would not be inserted.  We still need a type_id.    Consider the example of a 'SNP' page where the feature.type_id = 'genetic_marker', the featureprop.type_id = 'type' and the featureprop.value = 'SNP'.  You could define this content type, but if you wanted to manually add a new record you could not because the type_id would not be set to 'genetic_marker' and the prop record would not get inserted.   This PR fixes both of these problems.

This PR also includes a major overhaul of the content type creation form.   To make this PR work I needed to add some new form elements and that made the form overwhelming to look at.  The form is now a multi-step form (see screenshots below).  This should make it easier for end-users.

This PR also includes an update to the user_guide documents reflecting the changes made by this PR.

Note: This PR also includes tripal_chado_update_7331() to add a new column to the chado_bundle table for storing the type_id for the base table if need be.


## Testing?




1)  First, pull the PR branch.
2)  Run the update function using this command rather than updatedb (that way you can run updatedb on the 7.x-3.x branch once all update hooks have been merged in ... there are two other update hooks in two different open PRs).
  ```
  drush php-eval 'module_load_include("install", "tripal_chado", "tripal_chado"); 
tripal_chado_update_7331()'
  ```
3)  Rebuild the documentation:
  ```
  cd docs
  make html
  ```
4) Follow the two examples in the updated documentation to create two new content types.  One using a traditional method and the other using the prop table.  This will also give you a chance to review the updated docs.
5)  Try to create new content for both the new genetic_marker and SNP content types. You should be able to do so using the web interface and all should be inserted with no issues.  
6)  If you run this query you should see in the database that for this new feature added you will also have a property record added and the featureprop.type_id should resolve to the 'rdfs:type' term and the value should be 'SNP'.  The feature.type_id should be 'genetic_marker:
  ``` sql
  SELECT F.name, F.uniquename, CVT.name, CVP.name, CVTP.name, FP.value
  FROM featureprop FP
    INNER JOIN feature F on F.feature_id = FP.feature_id
    INNER JOIN cvterm CVT on F.type_id = CVT.cvterm_id
    INNER JOIN cvterm CVTP on FP.type_id = CVTP.cvterm_id
    INNER JOIN cv CVP on CVTP.cv_id = CVP.cv_id
  WHERE CVT.name = 'genetic_marker'
  ```
7) Please feel free to create other content types using different combinations of settings.  All should work, and all new content types should allow you to add new entities via the web interface without problems.

## Screenshots (if appropriate):
Sequence of form steps for creating a 'genetic_marker' content type:

![image](https://user-images.githubusercontent.com/1719352/44842280-473aa700-abfa-11e8-8ce3-3a8ce6d768d5.png)
![image](https://user-images.githubusercontent.com/1719352/44842289-4ace2e00-abfa-11e8-9b40-65185baa8acf.png)
![image](https://user-images.githubusercontent.com/1719352/44842292-4d308800-abfa-11e8-9971-cf9a3950a774.png)
![image](https://user-images.githubusercontent.com/1719352/44842298-50c40f00-abfa-11e8-82fc-ae2c570e39e3.png)
![image](https://user-images.githubusercontent.com/1719352/44842304-54f02c80-abfa-11e8-849a-eb93ab0ac231.png)

Creating a SNP page:
![image](https://user-images.githubusercontent.com/1719352/44843225-a994a700-abfc-11e8-82ea-ba0b6979420d.png)


## Additional Notes (if any):
